### DISCUSSION TITLE
Restart the treesitter parser for a buffer

### DISCUSSION THREAD

**Discussion by Dronakurl** [nvim-treesitter/nvim-treesitter#5684] [Q&A]
Hi, is it possible to restart treesitter on a buffer?
I want to set the injections query with a command.
 
This is the command I send: 
``` lua
require("vim.treesitter.query").set("python", "injections", "")
```
However, I need to reload the buffer, before I see any changes to the syntax highlighting.
Is it possible to do this in place? I tried  
 
``` lua
require("vim.treesitter").stop()
require("vim.treesitter").start()
```
and something like this: 
``` lua
vim.api.nvim_buf_clear_namespace(0,-1,0,-1)
require("nvim-treesitter.query").invalidate_query_cache()
```
but this has no effect. The same applies when I try to set the query for highlights.

**Comment by clason** [comment]
it's just `:lua vim.treesitter.stop()` etc.

**Reply by Dronakurl** [reply]
It is the same as  `require("vim.treesitter").stop()` and it has no effect. 
Strangely, this only concerns injections, not highlights. 
 
Try this minimal example. Define this user command, which should inject javascript on all text within tags and should set the tag names to be highlighted in bold: 
```lua
vim.api.nvim_create_user_command("KTestTS", function()
  vim.treesitter.query.set("html", "injections", '((text) @injection.content (#set! injection.language "javascript"))')
  vim.treesitter.query.set("html", "highlights", "((tag_name) @text.title)")
  vim.treesitter.stop()
  vim.treesitter.start()
end, {})
``` 
 
Now, apply this comand to a html file:
```html
<body>
  <style>
    body {
      background-color: white;
    }
  </style>
  <test> const a=23 </test>
</body>
``` 
The highlights (bold tags) are changed immediately, but the injections are not set. Only after reloading the buffer (with ":e", the injections become effective)

**Reply by clason** [reply]
Yes, injections are only reparsed on changed text. That is a (necessary) performance optimization. 

Just get the current parser and call parse(true) manually.

**Reply by Dronakurl** [reply]
Thank you for the hint, could you please help me out with a line of code which does that? I couldn't figure it out.

**Reply by clason** [reply]
```lua
local parser = vim.treesitter.get_parser()
parser:parse(true)
```

**Reply by Dronakurl** [reply]
Thanks again, but this still does nothing with the injections.
I also tried
```lua
parser:destroy()
parser:invalidate(true)
```
.. ,but to no effect (also in combination with stop() and start()).
Any other ideas? I couldn't find the code that is executed when a new buffer is attached.
